src/charm.py

```diff
@@ -227,165 +229,45 @@ def _on_update_status(self, event: EventBase) -> None:
-        self.kafka_config.set_zk_jaas_config()
-        self.kafka_config.set_server_properties()
-        self.kafka_config.set_client_properties()
-
-        # start kafka service
-        self.snap.start_snap_service()
-        logger.info("Kafka snap started")
-
-        # check for connection
-        self._on_update_status(event)
-
-        # only log once on successful 'on-start' run
-        if isinstance(self.unit.status, ActiveStatus):
-            logger.info(f'Broker {self.unit.name.split("/")[1]} connected')
-
-    def _on_config_changed(self, event: EventBase) -> None:
-        """Generic handler for most `config_changed` events across relations."""
-        # only overwrite properties if service is already active
-        if not self.healthy or not self.upgrade.idle:
-            event.defer()
-            return
-
-        # Load current properties set in the charm workload
-        properties = safe_get_file(self.kafka_config.server_properties_filepath)
-        if not properties:
-            # Event fired before charm has properly started
-            event.defer()
-            return
-
-        if set(properties) ^ set(self.kafka_config.server_properties):
-            logger.info(
-                (
-                    f'Broker {self.unit.name.split("/")[1]} updating config - '
-                    f"OLD PROPERTIES = {set(properties) - set(self.kafka_config.server_properties)}, "
-                    f"NEW PROPERTIES = {set(self.kafka_config.server_properties) - set(properties)}"
-                )
-            )
-            self.kafka_config.set_server_properties()
-
-            self.kafka_config.set_environment()
-
-            if isinstance(event, StorageEvent):  # to get new storages
-                self.on[f"{self.restart.name}"].acquire_lock.emit(
-                    callback_override="_disable_enable_restart"
-                )
-            else:
-                logger.info("Acquiring lock from _on_config_changed...")
-                self.on[f"{self.restart.name}"].acquire_lock.emit()
-
-        # update client_properties whenever possible
-        self.kafka_config.set_client_properties()
-
-        # If Kafka is related to client charms, update their information.
-        if self.model.relations.get(REL_NAME, None) and self.unit.is_leader():
-            self.provider.update_connection_info()
-
     def _restart(self, event: EventBase) -> None:
         """Handler for `rolling_ops` restart events."""
         # only attempt restart if service is already active
         if not self.healthy:
             event.defer()
             return
 
-        self.snap.restart_snap_service()
+        self.workload.restart()
 
         # FIXME: This logic should be improved as part of ticket DPE-3155
         # For more information, please refer to https://warthogs.atlassian.net/browse/DPE-3155
-        time.sleep(20.0)
+        time.sleep(10.0)
 
-        if self.healthy:
+        if self.workload.active():
```

Comment:
**thought/note:** Non-blocking.

I'm sure there was a good reason not to, but why don't we do `broker_active` here?
I guess to address the `FIXME` above it, there's something we can do